cli/stats_helpers.go

```diff
@@ -0,0 +1,283 @@
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/alibaba/pouch/apis/types"
+	"github.com/alibaba/pouch/client"
+
+	"github.com/docker/go-units"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// StatsEntry represents the statistics data collected from a container
+type StatsEntry struct {
+	container        string
+	name             string
+	id               string
+	cpuPercentage    float64
+	memory           float64
+	memoryLimit      float64
+	memoryPercentage float64
+	networkRx        float64
+	networkTx        float64
+	blockRead        float64
+	blockWrite       float64
+	pidsCurrent      uint64
+	err              error
+}
+
+// StatsEntryWithLock represents an entity to store containers statistics synchronously
+type StatsEntryWithLock struct {
+	mutex sync.Mutex
+	StatsEntry
+}
+
+// Name return the name of container
+func (s StatsEntry) Name() string {
+	return s.name
+}
+
+// ID return the id of container
+func (s StatsEntry) ID() string {
+	var shortLen = 12
+	if len(s.id) > shortLen {
+		return s.id[:shortLen]
+	}
+	return s.id
+}
+
+// CPUPerc return the cpu usage percentage
+func (s StatsEntry) CPUPerc() string {
+	if s.err != nil {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%.2f%%", s.cpuPercentage)
+}
+
+// MemUsage return memory usage
+func (s StatsEntry) MemUsage() string {
+	if s.err != nil {
+		return fmt.Sprintf("-- / --")
+	}
+
+	return fmt.Sprintf("%s / %s", units.BytesSize(s.memory), units.BytesSize(s.memoryLimit))
+}
+
+// MemPerc return memory percentage
+func (s StatsEntry) MemPerc() string {
+	if s.err != nil {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%.2f%%", s.memoryPercentage)
+}
+
+// NetIO return net IO usage of container
+func (s StatsEntry) NetIO() string {
+	if s.err != nil {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%s / %s", units.HumanSizeWithPrecision(s.networkRx, 3), units.HumanSizeWithPrecision(s.networkTx, 3))
+}
+
+// BlockIO return block IO usage of container
+func (s StatsEntry) BlockIO() string {
+	if s.err != nil {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%s / %s", units.HumanSizeWithPrecision(s.blockRead, 3), units.HumanSizeWithPrecision(s.blockWrite, 3))
+}
+
+// PIDs return current pid of container
+func (s StatsEntry) PIDs() string {
+	if s.err != nil {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%d", s.pidsCurrent)
+}
+
+// GetStatsEntry return the StatsEntry of StatsEntryWithLock
+func (s *StatsEntryWithLock) GetStatsEntry() StatsEntry {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.StatsEntry
+}
+
+// SetError will set error of StatsEntryWithLock
+func (s *StatsEntryWithLock) SetError(err error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.err = err
+}
+
+// GetError will return the error string of StatsEntryWithLock
+func (s *StatsEntryWithLock) GetError() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return fmt.Errorf("error stats container %s, err = %v", s.id, s.err.Error())
```

Comment:
`GetError()` always != nil here. 
First check whether s.err == nil, if does, return nil, else return fmt.Errorf(xxx)

Reply:
my bad, fix it.